File: LilysHomework/Go/lylys_solver.go
package main

import (
    "bufio"
    "fmt"
    "io"
    "os"
    "strconv"
    "strings"
    "sort"
)

/*
 * Complete the 'lilysHomework' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts INTEGER_ARRAY arr as parameter.
 */

func min(x, y int32) int32 {
    if x < y {
        return x
    }
    return y
}
func lilysHomework(arr []int32) int32 {
    // Write your code here
    arr_org := make([]int32, len(arr))
    copy(arr_org, arr)
    // fmt.Println(arr_org)
    sort.Slice(arr, func(i, j int) bool { return arr[i] < arr[j] })
    arr_rev := make([]int32, len(arr))
    copy(arr_rev, arr)
    sort.Slice(arr_rev, func(i, j int) bool {
      return arr_rev[j] < arr_rev[i]
   })
    fmt.Println("arr_org, arr, arr_rev", arr_org, arr, arr_rev)
    mapArr := make(map[int32]int32)
    for i, mapAr := range arr_org{
        mapArr[mapAr] = int32(i)
    }
    mapArrDec := make(map[int32]int32)
    for k, v := range mapArr{
         mapArrDec[k] = v       
    }
    fmt.Println("mapArr", mapArr, mapArrDec)
    
    arr_org_inc := make([]int32, len(arr_org))
    copy(arr_org_inc, arr_org)
    arr_org_dec := make([]int32, len(arr_org))
    copy(arr_org_dec, arr_org)
    var ret_inc int32 = 0
    var ret_dec int32 = 0
    for i:=0; i<len(arr_org_inc); i++ {
        if arr[i] != arr_org_inc[i]{
            ret_inc++
            need_val := arr[i]
            swap_val := arr_org_inc[i]
            arr_org_inc[i] = need_val
            arr_org_inc[mapArr[need_val]] = swap_val
            idx_need_val := mapArr[need_val]
            mapArr[need_val] = int32(i)
            mapArr[swap_val] = idx_need_val   
        }
    }
    
    for j:=0; j<len(arr_org_dec); j++ {
        if arr_rev[j] != arr_org_dec[j]{
            ret_dec++
            need_val := arr_rev[j]
            swap_val := arr_org_dec[j]
            arr_org_dec[j] = need_val
            arr_org_dec[mapArrDec[need_val]] = swap_val
            idx_need_val := mapArrDec[need_val]
            mapArrDec[need_val] = int32(j)
            mapArrDec[swap_val] = idx_need_val   
        }
    } 
    
    return min(ret_inc, ret_dec)

}

func main() {
    reader := bufio.NewReaderSize(os.Stdin, 16 * 1024 * 1024)

    stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
    checkError(err)

    defer stdout.Close()

    writer := bufio.NewWriterSize(stdout, 16 * 1024 * 1024)

    nTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
    checkError(err)
    n := int32(nTemp)

    arrTemp := strings.Split(strings.TrimSpace(readLine(reader)), " ")

    var arr []int32

    for i := 0; i < int(n); i++ {
        arrItemTemp, err := strconv.ParseInt(arrTemp[i], 10, 64)
        checkError(err)
        arrItem := int32(arrItemTemp)
        arr = append(arr, arrItem)
    }

    result := lilysHomework(arr)

    fmt.Fprintf(writer, "%d\n", result)

    writer.Flush()
}

func readLine(reader *bufio.Reader) string {
    str, _, err := reader.ReadLine()
    if err == io.EOF {
        return ""
    }

    return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
    if err != nil {
        panic(err)
    }
}

File: LilysHomework/Go/lylys_solver.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);
vector<string> split(const string &);

/*
 * Complete the 'lilysHomework' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts INTEGER_ARRAY arr as parameter.
 */

int lilysHomework(vector<int> arr) {
    vector<int> arr_org(arr);
    vector<int> arr_inc(arr);
    vector<int> arr_dec(arr);
    vector<int> sorted_arr_inc(arr);
    sort(sorted_arr_inc.begin(), sorted_arr_inc.end());
    vector<int> sorted_arr_dec(sorted_arr_inc);
    reverse(sorted_arr_dec.begin(), sorted_arr_dec.end());
        
    unordered_map<int, int> mapArr;
    for (int i=0; i < arr_org.size(); i++)
    {
        mapArr[arr_org[i]] = i;
    }
    unordered_map<int ,int> mapArrDec = mapArr;
    int ret_inc = 0;
    int ret_dec = 0;
    for (int i=0; i < arr_inc.size(); i++)
    {
        if(sorted_arr_inc[i] != arr_inc[i]){
            ret_inc++;
            int desire_val = sorted_arr_inc[i];
            int swap_val = arr_inc[i];
            int desire_val_idx = mapArr[desire_val];
            int swap_val_idx = i;
            arr_inc[swap_val_idx] = desire_val;
            arr_inc[desire_val_idx] = swap_val;
            mapArr[desire_val] = swap_val_idx;
            mapArr[swap_val] = desire_val_idx;
        }
    }
    
    for (int i=0; i < arr_dec.size(); i++)
    {
        if(sorted_arr_dec[i] != arr_dec[i]){
            ret_dec++;
            int desire_val = sorted_arr_dec[i];
            int swap_val = arr_dec[i];
            int desire_val_idx = mapArrDec[desire_val];
            int swap_val_idx = i;
            arr_dec[swap_val_idx] = desire_val;
            arr_dec[desire_val_idx] = swap_val;
            mapArrDec[desire_val] = swap_val_idx;
            mapArrDec[swap_val] = desire_val_idx;
        }
    }
    
    return min(ret_inc, ret_dec);
    
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string n_temp;
    getline(cin, n_temp);

    int n = stoi(ltrim(rtrim(n_temp)));

    string arr_temp_temp;
    getline(cin, arr_temp_temp);

    vector<string> arr_temp = split(rtrim(arr_temp_temp));

    vector<int> arr(n);

    for (int i = 0; i < n; i++) {
        int arr_item = stoi(arr_temp[i]);

        arr[i] = arr_item;
    }

    int result = lilysHomework(arr);

    fout << result << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}

vector<string> split(const string &str) {
    vector<string> tokens;

    string::size_type start = 0;
    string::size_type end = 0;

    while ((end = str.find(" ", start)) != string::npos) {
        tokens.push_back(str.substr(start, end - start));

        start = end + 1;
    }

    tokens.push_back(str.substr(start));

    return tokens;
}